### PR TITLE
feat: *.localhost subdomain gateway support with http proxy

### LINKED ISSUE #328: Mixed Content Warning when using `localhost` Gateway
This one is kind of _Rube Goldberg machine_, so I'll start with a bit of background information.

_[Mixed Content](https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content) warning_ happens when a page loaded over HTTPS contains a resource (eg. image) that  is loaded from insecure HTTP.  Browser vendors recognized that loading content over unencrypted HTTP from the box browser is running on is  a special case that provides similar security guarantees to HTTPS over the internet, so they whitelisted `127.0.0.1` and it no longer trigger the warning.

IPFS Companion defaults to using `localhost` instead of `127.0.0.1` since  #291 was closed. To be more specific, if user enters gateway URL that includes `127.0.0.1` it will be replaced with hostname `localhost`. 

The goal was to improve UX in Location Bar (`localhost` looks _much_ better than raw IP), but it came with unexpected side effect regarding [Mixed Content](https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content) whitelisting implemented in Firefox and Chrome.  It seems that browsers whitelisted `127.0.0.1` but not `localhost`, which still produces Mixed Content warning.

In our use case, the **Mixed Content Warning is triggered when there is an IPFS resource on HTTPS page and ipfs-companion's redirect loads it from Local Gateway that has `localhost` in its URL**.

----

_(I used [pls.watch](https://pls.watch/) as a PoC in examples below,  because it is a public HTTPS site that enables users to trigger Mixed Content Warning by loading external resources from HTTP URLs)_

#### Firefox: the green padlock

> 2017-12-09-125633_713x31_scrot

#### Firefox: grey padlock with warning

> 2017-12-09-125548_691x29_scrot

> 2017-12-09-125858_1147x1421_scrot

> 2017-12-09-125707_655x60_scrot

#### Chrome: no padlock, warning under  `(i)` icon

> 2017-12-09-125837_1145x1419_scrot

---- 

For now, we need to decide what is more important:  

* keeping human-readable `localhost` in Location Bar at a cost of signaling that use of local IPFS gateway is somehow "not secure"

or 

* keeping "the green padlock" at a cost of going back to `127.0.0.1` in location bar.

Personally, I think "keeping the green  padlock"  is more important and we should go back and default to  `127.0.0.1`. Rationale: we can work with browser vendors to whitelist `localhost` and switch back to it at a later time, when it does not cause Mixed Content Warning.  

I am usually biased towards security  over UX, so would love to hear what others think before doing any changes.

----

References:

- https://bugzilla.mozilla.org/show_bug.cgi?id=903966 (Firefox whitelists `127.0.0.1` since v55)
- https://chromium.googlesource.com/chromium/src.git/+/130ee686fa00b617bfc001ceb3bb49782da2cb4e (Chrome whitelis `127.0.0.1` since v53)

### LINKED ISSUE #736: Use Web UI version from IPFS Node
## Context

Every IPFS Node provides `:5001/webui` endpoint on the API port. It redirects to the latest 'blessed' CID that can be loaded from `:5001/ipfs/{cid}`. For security reasons only hardcoded CIDs can be loaded from API port.

Right now IPFS Companion ships with hardcoded version of Web UI, and that version can be bigger than the one at `:5001/webui`.

Companion can't load it from API port as it is not on the whitelist, so it loads it from the Gateway instead. That comes at a price of having two separate Origins and falling under CORS limitations. We don't want people to add CORS headers, as it would allow allow any website loaded from Gateway port to execute arbitrary requests on API port. Instead, we remove CORS check on webui path by supressing `Origin` header in requests made from `:8080/ipfs/{blessed_webui_cid}` (Gateway) to `:5001` (API). 
That way we keep CORS check from anything else on Gateway port.

Why we did this? It made it possible for us to iterate and ship new Web UI without waiting for new go-ipfs release (which took multiple months at the time).

## Proposed change

We are now in maintenance mode where regular go-ipfs releases are expected and webui itself is much more stable now.

It is time to:

- now: switch to `/webui` provided by go-ipfs (or js-ipfs) as the new default in Companion
  - PoC in PR: https://github.com/ipfs-shipyard/ipfs-companion/pull/737
- future: provide opt-in option for Companion users to load latest Web UI from `:5001/ipns/webui.ipfs.io` (requires ipns path to be added to API port whitelist in GO and JS)
  - PoC PR for Companion adding opt-in option: https://github.com/ipfs-shipyard/ipfs-companion/pull/738
    - PoC PR with discussion for go-ipfs: https://github.com/ipfs-shipyard/ipfs-companion/issues/736
    - PoC PR with discussion for js-ipfs: https://github.com/ipfs/js-ipfs/pull/2281

cc @olizilla  (https://github.com/ipfs-shipyard/ipfs-webui/issues/1038#issuecomment-494000620)

### PR DESCRIPTION
**Pull request by lidel**
This PR adds support for subdomain gateways introduced in go-ipfs v0.5.0-dev (https://github.com/ipfs/go-ipfs/pull/6096), specifically, one running at `*.localhost` subdomains.

### Um.. what? 

**TL;DR** we will change the URL present in the address bar when user loads content from the local node (ipfs-desktop or standalone go-ipfs):

```
http://127.0.0.1:8080/ipns/en.wikipedia-on-ipfs.org/wiki/
```

to:

```
http://en.wikipedia-on-ipfs.org.ipns.localhost:8080/ipns/wiki/
```

#### Why?

- **localhost subdomain gateway provides unique Origin per content root**
  - this is extremely important, see
    - historical discussion about URI and Origin
      -  https://github.com/ipfs/in-web-browsers/issues/6
      - https://github.com/ipfs/in-web-browsers/issues?q=label%3AOrigin
    - in-depth analysis of how to unify the way we do redirects without compromising website isolation  https://github.com/ipfs-shipyard/ipfs-companion/issues/667
- better UX on DNSLink websites
- DNSLink websites get proper Origin isolation per content root (https://github.com/ipfs-shipyard/ipfs-companion/issues/667)
- we can enable redirect of subdomain gateways  (https://github.com/ipfs/in-web-browsers/issues/89, https://github.com/ipfs-shipyard/ipfs-companion/issues/667)

### How?

> Disclaimer: below is a draft of what may end up a blogpost after go-ipfs 0.5 ships. WIP. 

They key challenge was to ensure *.localhost DNS names resolve to 127.0.0.1 on all platforms, specifically Linux distribution which ship with very basic DNS resolver that returns "not found" when trying to resolve localhost subdomains. 

Browsers are moving in the direction of hardcoding `localhost` names to loopback interface, but we are not there yet. We need to solve this problem with tools at hand.

We do that by setting up HTTP Gateway port of local go-ipfs to act as HTTP Proxy. This removes DNS lookup step from the browser, and go-ipfs ships with implicit support for subdomain gateway when request comes with "Host: <cid>.ipfs.localhost:8080" or
similar. 

We register HTTP proxy using Firefox and Chromium-specific APIs, but the  end result is the same. When enables, default gateway uses 'localhost'  hostname (subdomain gateway) instead of '127.0.0.1' (path gateway)  and every path-pased request gets redirected to subdomain by go-ipfs itself, which decreases complexity on browser extension side.

#### TODO

This is work in progress:

- [x] support enabling/disabling HTTP proxy
  - [x] Firefox
  - [x] Chromium
- [x] make `localhost` the default
- [x] update old configs from `127.0.0.1` to `localhost`
- [x] fix context actions on *.localhost
  - [x] release `is-ipfs` with  `isIPFS.dnslinkSubdomain(url)`: https://github.com/ipfs/is-ipfs/pull/32
- [x] redirect from `127.0.0.1` to `localhost`  when proxy is enabled (unify user experience)
- [x] enable redirect of subdomain gateways (like`*.dweb.link`) to `*.locahost`
- [x] fix toggle per site on `<fqdn>.ipns.localhost`
- [x] fix _Copy Public Gateway URL_ on `<fqdn>.ipns.localhost`
- [x] ship a beta (before merging this PR)
- [x] confirm default settings work fine with go-ipfs 0.4.23
- [x] confirm default settings work fine with go-ipfs 0.5.0-dev
- [x] confirm default settings work fine in Brave
- [x] fix / add tests
  - [x] https://ipfs.io/ipfs/QmbcBPAwCDxRMB1Qe7CRQmxdrTSkxKwM9y6rZw2FjGtbsb/%3Ffilename=test.jpg can be opened from local gw
  - [x] do not redirect non-main-frame requests to local subdomain, so embedding images works in firefox
  - [x] do not proxy all `localhost`  ports, only the one of current gateway
  - [x] state.enabled = false should disable proxy

### caveats 

- [match pattern](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns) used for registering proxy handler in Firefox uses `localhost` name without port. This is by design, match patterns do not seem to support ports and are executed on all of them. We need to check port inside of handler, and return `direct` proxy route if it does not match
- proxy situation is temporary and will change over time
  - Chromium does not need proxy, they hardcode `localhost` and `*.localhost` to loopback IPs
  - Firefox needs proxy until https://bugzilla.mozilla.org/show_bug.cgi?id=1220810 lands

### Related

- Migration to CIDv1 (default base32) https://github.com/ipfs/ipfs/issues/337
- CID as a subdomain https://github.com/ipfs/in-web-browsers/issues/89
- Redirect of CID-in-subdomain and DNSLink websites: https://github.com/ipfs-shipyard/ipfs-companion/issues/667
- Subdomain Gateway https://github.com/ipfs/go-ipfs/issues/6498
  - Act as HTTP PROXY for `http://<cidv1b32>.ipfs.localhost`: https://github.com/ipfs/go-ipfs/issues/5982
  - PR: subdomain gateway via http proxy in go-ipfs: https://github.com/ipfs/go-ipfs/pull/6096
- Closes #328
- Closes #736

### CODE REVIEW

**Comment by lidel** [comment]
Surfacing some notes on additional fixes that had to be applies to keep subresource redirects working on HTTPS websites:

> Firefox 74 does not mark 	`*.localhost subdomains`	 as Secure Context yet
> (https://bugzilla.mozilla.org/show_bug.cgi?id=1220810#c23) so we can't redirect
> there when we have IPFS resource embedded on HTTPS page (eg.  image loaded from
> a public gateway) because that would cause mixed-content warning and
> subresource would fail to load.  Given the fact that `localhost/ipfs/*` provided
> by go-ipfs 0.5+ returns a redirect to `*.ipfs.localhost` subdomain we need to
> check requests for subresources, and manually replace `localhost` hostname with
> `127.0.0.1` (IP is hardcoded as Secure Context in Firefox). The need for this
> workaround can be revisited when Firefox closes mentioned bug.
> 
> Chromium 80 seems to force HTTPS in the final URL (after all redirects) so
> `https://*.localhost` fails. This needs additional research (could be a bug in
> Chromium). For now we reuse the same workaround as Firefox.
> 
> To unify use of `127.0.0.1` and `localhost` in address bar (eg. when user opens an
> image in a new tab etc) when Subdomain Proxy is enabled we normalize address
> bar requests made to the local gateway and replace raw IP with 'localhost'
> hostname to take advantage of subdomain redirect provided by go-ipfs >= 0.5

----

In other news:

- Added a bunch of tests, all green now. 
- In the meantime I noticed Chrome does not require proxy mode because they already hardcode `localhost` to raw IP. I am considering removing `proxy` permission from Chromium package as it may trigger yet another manual review at Chrome Web Store. 
  - Will keep the code commented out in case we need to enable it in the future for some reason.

File: add-on/_locales/en/messages.json [hunk 1]
```diff
@@ -279,6 +279,14 @@
...
     "description": "An option description on the Preferences screen (option_useCustomGateway_description)"
   },
+  "option_useSubdomains_title": {
+    "message": "Use Subdomains",
+    "description": "An option title on the Preferences screen (option_useSubdomains_title)"
+  },
+  "option_useSubdomains_description": {
+    "message": "Isolate content roots from each other by loading them from subdomains at *.localhost and creating a unique Origin for each CID, IPNS or DNSLink record. Requires a local go-ipfs 0.5.0 or later.",
+    "description": "An option description on the Preferences screen (option_useSubdomains_description)"
+  },
```
Comment (lidel) [inline add-on/_locales/en/messages.json:289]: On _Preferences_ screen the above labels look like this:



@jessicaschilling @autonome I suspect this could be phrased in a more human way, but this is the best I could do for next Beta release.  Suggestions welcome before we ship to Stable.

This is a description of a toggle that lets people to switch to old path-based gateway on `127.0.0.1` if they choose to do so for some reason. (disabling it stops HTTP proxy and swaps Custom Gateway URL to use raw IP)

**Comment by lidel** [comment]
Shipped this in a new Beta: [v2.10.0.902](https://github.com/ipfs-shipyard/ipfs-companion/releases/tag/v2.10.0.902).  Will play with it during the weekend.

I do not expect big changes apart from tweaking labels mentioned in previous comment.

File: test/functional/lib/ipfs-request-workarounds.test.js [hunk 2]
```diff
@@ -146,65 +148,23 @@ describe('modifyRequest processing', function () {
...
-      const response = Object.assign({}, request)
-      delete response.requestHeaders
-      response.responseHeaders = []
-      const responseHeaders = modifyRequest.onHeadersReceived(response).responseHeaders
-      const corsHeader = { name: 'Access-Control-Allow-Origin', value: state.gwURL.origin }
-      const acahHeader = { name: 'Access-Control-Allow-Headers', value: acrhHeader.value } // expect value restored from preflight
-      expect(responseHeaders).to.deep.include(corsHeader)
-      expect(responseHeaders).to.deep.include(acahHeader)
-
+      expect(modifyRequest.onBeforeSendHeaders(request).requestHeaders).to.include(nullOriginHeader)
```
Comment (lidel) [inline test/functional/lib/ipfs-request-workarounds.test.js:167]: I've added test to ensure Companion does not remove `Origin: null`

File: test/functional/lib/ipfs-request-workarounds.test.js [hunk 3]
```diff
@@ -146,65 +148,23 @@ describe('modifyRequest processing', function () {
   })
 
   describe('a request to <apiURL>/api/v0/ with Origin=null', function () {
-    it('should remove Origin header ', async function () {
-      // set vendor-specific Origin for WebExtension context
+    it('should keep the "Origin: null" header ', async function () {
+      // Presence of Origin header is important as it protects API from XSS via sandboxed iframe
+      // NOTE: Chromium <72 was setting this header in requests sent by browser extension,
+      // but they fixed it since then.
```
Comment (lidel) [inline test/functional/lib/ipfs-request-workarounds.test.js:154]: Removed unused code related to loading webui from gateway port without setting CORS and added test to ensure Companion does not remove `Origin: null` 

cc https://github.com/ipfs/go-ipfs-cmds/pull/190

**Comment by lidel** [comment]
After feedback from two beta releases I feel this is stable enough to merge to master and ship beta+stable. 

We need this not only for subdomains, but  also to ensure POST-only API (https://github.com/ipfs/go-ipfs/pull/7097) scheduled for go-ipfs 0.5 does break DNSLink lookup and ipfs-webui.

**Comment by lidel** [comment]
Released in Beta: https://github.com/ipfs-shipyard/ipfs-companion/releases/tag/v2.11.0.904
Scheduled to ship to stable as v2.11.0
